### ISSUE TITLE
Is there a way to disable the trigger button while a build is ongoing?

### ISSUE THREAD

**Issue by s-afroz** [tomasbjerre/pull-request-notifier-for-bitbucket#104]
We have an issue of developers clicking on the button repeatedly, since there is a delay between pressing the button, and getting the build notification displayed on the pull request. Sometimes, the build job waits on Jenkins for a build server to be available. Being unaware of this, the pull request originator keeps pressing the trigger button, thus starting multiple builds from the same commit.

I was wondering if it is possible to disable the trigger button as soon as it is pressed and while a build is ongoing with the latest commit on the pull request.

For now, we hid this button from regular users and reserved only for system admins. Developers now only can trigger a job by other means (open/re-open).

**Comment by tomasbjerre** [comment]
Perhaps you can use the ${INJECTION_URL_VALUE} variable for that.
A quick googling says that this URL will give you the commit that is being built:
http://jenkins/job/myjob/../api/json?tree=changeSet[*[*]]
So perhaps use that as the *injection url*.
If the filter string includes ${INJECTION_URL_VALUE}, the filter regexp may include ${PULL_REQUEST_FROM_HASH} .

But I guess you may have the same delay here. Does you Jenkins build job have a sleep period before build starts? 

![sleep](https://cloud.githubusercontent.com/assets/439571/13709219/008edbb8-e7b3-11e5-8d3d-5e254df2b7cb.png)

This is Swedish now, sorry, but the value that is 0 in the screenshot is 5 by default. Perhaps that will help you make the build start instantly.

Tell me how that goes =)

**Comment by s-afroz** [comment]
No, there is no delay. Due to limited resources, we allocate a limited # of build servers for the pull request builds. If there is heavy load for a repo, then, an incoming build request may have to wait until a build server is available.

**Comment by fredrikand** [comment]
While I have not tried the approach described by @tomasbjerre , I can propose another way.
I use the Build Flow DSL build type and use the following code to stop duplicated builds:

```
def isMe(it) {
  //out.println "Is it me?"
  //out.println it == build
  return it == build
}

//
// Find out if we should abort due to same commit already building
//
def job = Jenkins.getInstance().getItemByFullName(build.environment.get('JOB_NAME'))
def should_abort=false
// Check if a job like this one is already running
job.builds.findAll { it.isBuilding() && ! isMe(it) }.each { 
  
  if ( it.getEnvironment()['PULL_REQUEST_FROM_HASH'] == build.environment.get('PULL_REQUEST_FROM_HASH') ) {
    if ( it.getEnvironment()['BUILD_NUMBER'].toInteger() < build.environment.get('BUILD_NUMBER').toInteger() ) {
      out.println "Abort since same commit is already running in this job:"
      out.println it
      should_abort=true
    }
  }
}
if ( should_abort ) {
  return
}
```

**Comment by s-afroz** [comment]
Thank you for sharing your method. BTW, wouldn't an aborted build report a failed build to the stash pull request? We enforce a successful build in order to merge. In this case, even if the earlier build passes, the later failed build will prohibit developer from merging the PR (may not be a bad idea in the long run, making them not to click trigger button back-to-back for the same commit).

Would be really great if the pull request notifier could grey out the trigger-build if a build request has already been sent to the Jenkins server for the same commit  ;)

**Comment by fredrikand** [comment]
Yes, it would be super if there was a status-interaction with for example Jenkins.

I am not using the Stash/Bitbucket Notifier plugin for the triggered job, but instead call the Bitbucket REST API with status INPROGRESS after the job script passed the should_abort point. Also, reporting the FAILURE or SUCCESS status when that information is known.
Should mention that the BuildFlow dsl job is orchestrating many other jobs and in those I am using the Stash Notifier plugin (because it is simple and works well in those cases). To prevent a PR to get merged before all jobs are built (we almost always have a queue of jobs), the orchestrating job is waiting for all other jobs before reporting SUCCESS and in this way is acting as a blocker until all other jobs has been built.
